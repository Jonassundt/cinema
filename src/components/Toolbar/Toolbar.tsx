import React, { useContext, useState, useEffect } from 'react';

import posterImages from "../Images";

import Parameters from "../../Parameters";

import styles from './toolbar.module.css';

function Toolbar() {
    /* Hooks to handle the parameters given in the toolbar element.
    *The parameters controls how the website is to be displayed.
    * Parameters control brightness, volume, animations, colours, fullscreenmode, slideshowmode and if the user wishes to only see his/her favourite posters.
    */
    const { params, setParams } = useContext(Parameters);
    const setLight = (light: Number) => { sessionStorage.setItem("light", (light).toString()); setParams({ ...params, light }) };
    const setVolume = (volume: Number) => setParams({ ...params, volume });
    const setAnimation = () => { sessionStorage.setItem("animation", (!params.animation).toString()); setParams({ ...params, animation: !params.animation }) };
    const setColor = () => { sessionStorage.setItem("color", (!params.color).toString()); setParams({ ...params, color: !params.color }) };
    const setFullscreen = () => { sessionStorage.setItem("fullscreen", (!params.fullscreen).toString()); setParams({ ...params, fullscreen: !params.fullscreen }) };
    const toggleSlideshow = () => setParams({ ...params, slideshow: !params.slideshow });
    const toggleFavorites = () => {
        sessionStorage.setItem("showFavorites", (!params.showFavorites).toString());
        window.scroll({
            top: 4000,
            behavior: 'smooth'
        }); setParams({ ...params, showFavorites: !params.showFavorites })
    };
    const [slideShowIntervalId, setSlideShowIntervalId] = useState(0);
    const [slideshowBar, setSlideshowBar] = useState(0);
    const setFavorite = () => setParams({ ...params, favorites: toggleIndexFavorite() });
    const setMood = () => { sessionStorage.setItem("mood", (!params.mood).toString()); setParams({ ...params, mood: !params.mood }) };

    const toggleIndexFavorite = () => { //Legge til movieIndex hvis den eksisterer i params.favorites,
        // Eller fjerne den fra lista om den allerede eksisterer.
        let filteredArray: number[] = [];
        if (params.favorites.includes(params.posterIndex)) { //Remove if movie already in favorites
            filteredArray = params.favorites.filter(movie => movie !== params.posterIndex);
        }
        else { //Hvis den ikke er i lista, gjÃ¸r den til favoritt.
            filteredArray = params.favorites;
            filteredArray.push(params.posterIndex);
        }
        localStorage.setItem("favorites", JSON.stringify(filteredArray));
        return filteredArray;
    }

    const slideshow = () => {
        //Method to initiate slideshow mode
        if (!params.slideshow) {
            const id = setInterval(() => setSlideshowBar(prev => (prev + 0.1) % 15), 100)
            setSlideShowIntervalId(parseInt(id.toString()));
        }
        else {
            clearInterval(slideShowIntervalId);
            setSlideShowIntervalId(0);
            setSlideshowBar(0)
        }
        toggleSlideshow();
    }

    useEffect(() => {
        if (!params.slideshow) {
            clearInterval(slideShowIntervalId);
            setSlideShowIntervalId(0);
            setSlideshowBar(0)
        }

    }, [params, slideShowIntervalId])

    if (params.loading) {
        return (
            <div className={params.fullscreen ? styles.ToolbarFullscreen : styles.Toolbar}>
            </div>
        )
    }

    return (
        <div style={{ display: "unset" }}>
            <div className={params.fullscreen ? styles.ToolbarFullscreen : styles.Toolbar}>
                {/* Lyssetting */}
                <div>
                    <img className={styles.Icon} src={"/icons/brightness.svg"} alt="" />
                    <div className={styles.Group}>
                        <div onClick={() => params.light < 77 && setLight(params.light + 4)}>+</div>
                        <div>{(params.light - 40) * 2.5}</div>
                        <div onClick={() => params.light > 40 && setLight(params.light - 4)}>-</div>
                    </div>
                </div>
                {/* Volum */}
                <div>
                    <img className={styles.Icon} src={"/icons/volume.svg"} alt="" />
                    <div className={styles.Group}>
                        <div onClick={() => params.volume < 10 && setVolume(params.volume + 1)}>+</div>
                        <div>{params.volume * 10}</div>
                        <div onClick={() => params.volume > 0 && setVolume(params.volume - 1)}>-</div>
                    </div>
                </div>
                {/* Animasjoner */}
                <div className={styles.Button} onClick={setAnimation}>Animasjon</div>
                {/* Sort-hvitt */}
                <div className={styles.Button} onClick={setColor}>Fargefilm</div>
                {/* Stemning */}
                <div className={styles.Button} onClick={setMood}>Stemning</div>
                {/* Fullskjerm */}
                <div className={styles.Button} onClick={setFullscreen}>{params.fullscreen ? "Vanlig skjerm" : "Fullskjerm"}</div>
                {/* Slideshow */}
                <div className={styles.ButtonBar} onClick={slideshow}>
                    <div className={styles.SlideBar} style={{ width: `${(slideshowBar / 15) * 100}%` }}>
                        <div className={styles.SlideText}>{!params.slideshow ? "Rullering" :
                            <div>Neste: <strong>{posterImages[(params.posterIndex + 1) % posterImages.length].props.name}</strong></div>}
                        </div>
                    </div>
                </div>
                {/* Favoritt */}
                <div className={styles.Row}>
                    <div className={styles.Toggle} onClick={setFavorite}>{params.favorites.includes(params.posterIndex) ? "Fjern" : "Lagre"} favoritt</div>
                    {/* Favoritter */}
                    {params.fullscreen ? null :
                        <div className={styles.Toggle} onClick={toggleFavorites}>{params.showFavorites ? "Se alle" : "Se favoritter"}</div>}
                </div>
            </div>
            <div className={styles.SpotRow} style={{ opacity: params.light < 55 ? 0 : 1 }}>
                <div className={styles.Spotlight} />
                <div className={styles.Spotlight} />
            </div>
        </div>
    );
}

export default Toolbar;
