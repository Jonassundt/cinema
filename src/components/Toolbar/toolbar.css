.Toolbar {
    padding: 40px;
    background-color: rgba(77, 77, 77, 0.5);
    border-radius: 20px;
    text-align: center;   
    margin-bottom: 100px; 
}