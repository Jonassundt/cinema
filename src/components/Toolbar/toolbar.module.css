  .Toolbar, .ToolbarFullscreen {
    padding: 20px;
    background: #171a36;
    border-radius: 20px;
    text-align: center;   
    margin-bottom: 60px; 
    display: flex;
    transition: 0.8s;
    z-index: 15;
    min-height: 113px;
}

.ToolbarFullscreen {
    position: sticky;
    top: 45px;
    margin-bottom: 40px;
    margin-top: 40px;
}

.Row {
    display: flex;
    margin-left: auto
}

.Group {
    font-size: 1.2em;
    font-weight: bold;
    background: #4d4d4d86;
    width: 50px;
    border-radius: 8px;
    cursor: pointer;
    user-select: none;
    margin: 0 10px;
}

.Group div:not(:nth-child(2)):hover {
    font-size: 1.2em;
    margin-top: -0.12em;
    margin-bottom: -0.12em;
}

.Toggle,
.ButtonBar,
.Button {
    background: #4d4d4d86;
    border-radius: 8px;
    cursor: pointer;
    user-select: none;
    margin: 0 10px;
    padding: 35px 25px;
    transition: 0.1s;
    min-width: 75px;
    height: 30px;
    margin-top: 10px;
}

.Toggle {
    background: rgba(133, 133, 133, 0.525);
    border-radius: 0 8px 8px 0;
    margin: 10px 10px 0 0;
    min-width: 50px;
    padding: 35px 20px;
}

.Toggle:first-child {
    border-radius: 8px 0 0 8px;
    margin: 10px 0 0 10px;
}

.Toggle:hover,
.Group:hover,
.ButtonBar:hover,
.Button:hover {
    background: rgba(119, 119, 119, 0.525);
}

.ButtonBar {
    padding: 0;
    overflow: hidden;
    height: 100px;
}

.Icon {
    width: 30px;
    height: 30px;
}

.SpotRow {
    display: flex;
    justify-content: space-between;
    padding: 0 180px;
}

.Spotlight {
    box-shadow: 0px 150px 170px 70px rgb(214, 150, 0);
    background: rgba(214, 178, 0, 0);
    height: 2px;
    width: 120px;
    border-radius: 20px;
    z-index: 10;
}

.SlideBar {
    background: #88888886;
    padding: 35px 0 50px 0;
    border-radius: 8px 0 0 8px;
}

.SlideText {
    width: 160px;
}