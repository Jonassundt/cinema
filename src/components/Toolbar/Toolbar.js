import React from 'react';

import './toolbar.css';

function Toolbar() {
    return (
        <div className="Toolbar">
            TODO: Add some basic settings and parameterization test
        </div>
    );
}

export default Toolbar;
