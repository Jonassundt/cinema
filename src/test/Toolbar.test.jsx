import React from 'react';
import renderer from 'react-test-renderer';
import Toolbar from '../components/Toolbar/Toolbar.tsx';

describe('testing the toolbar', () => {
    it("matches the snapshot", () => {
        const tree = renderer.create(<Toolbar />).toJSON();

        expect(tree).toMatchSnapshot();
    })
});