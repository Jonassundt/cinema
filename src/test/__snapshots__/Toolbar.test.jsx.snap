// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the toolbar matches the snapshot 1`] = `
<div
  style={
    Object {
      "display": "unset",
    }
  }
>
  <div
    className="Toolbar"
  >
    <div>
      <img
        alt=""
        className="Icon"
        src="/icons/brightness.svg"
      />
      <div
        className="Group"
      >
        <div
          onClick={[Function]}
        >
          +
        </div>
        <div>
          -100
        </div>
        <div
          onClick={[Function]}
        >
          -
        </div>
      </div>
    </div>
    <div>
      <img
        alt=""
        className="Icon"
        src="/icons/volume.svg"
      />
      <div
        className="Group"
      >
        <div
          onClick={[Function]}
        >
          +
        </div>
        <div>
          0
        </div>
        <div
          onClick={[Function]}
        >
          -
        </div>
      </div>
    </div>
    <div
      className="Button"
      onClick={[Function]}
    >
      Animasjon
    </div>
    <div
      className="Button"
      onClick={[Function]}
    >
      Fargefilm
    </div>
    <div
      className="Button"
      onClick={[Function]}
    >
      Stemning
    </div>
    <div
      className="Button"
      onClick={[Function]}
    >
      Fullskjerm
    </div>
    <div
      className="ButtonBar"
      onClick={[Function]}
    >
      <div
        className="SlideBar"
        style={
          Object {
            "width": "0%",
          }
        }
      >
        <div
          className="SlideText"
        >
          Rullering
        </div>
      </div>
    </div>
    <div
      className="Row"
    >
      <div
        className="Toggle"
        onClick={[Function]}
      >
        Fjern
         
        favoritt
      </div>
      <div
        className="Toggle"
        onClick={[Function]}
      >
        Se favoritter
      </div>
    </div>
  </div>
  <div
    className="SpotRow"
    style={
      Object {
        "opacity": 0,
      }
    }
  >
    <div
      className="Spotlight"
    />
    <div
      className="Spotlight"
    />
  </div>
</div>
`;
